rewriter/Data.cc

```diff
@@ -101,18 +167,35 @@ vector<ast::ExpressionPtr> Data::run(core::MutableContext ctx, ast::Assign *asgn
             symLoc = ctx.locAt(symLoc).adjust(ctx, 1, 0).offsets();
         }
 
-        sigArgs.emplace_back(ast::MK::Symbol(symLoc, name));
-        sigArgs.emplace_back(ast::MK::Constant(symLoc, core::Symbols::BasicObject()));
+        auto memberType = getMemberType(ctx, reliableSigParams, name, symLoc);
+        body.emplace_back(ast::MK::Sig0(symLoc, ASTUtil::dupType(memberType)));
+        body.emplace_back(ast::MK::SyntheticMethod0(symLoc, symLoc, name, ast::MK::RaiseUnimplemented(loc)));
 
-        auto argName = ast::MK::Local(symLoc, name);
-        newArgs.emplace_back(ast::MK::OptionalArg(symLoc, move(argName), ast::MK::Nil(symLoc)));
+        if (!(initialize.has_value() && initializeHasSig)) {
+            newSigArgs.emplace_back(ast::MK::Symbol(symLoc, name));
+            newSigArgs.emplace_back(ASTUtil::dupType(memberType));
+            auto argName = ast::MK::Local(symLoc, name);
+            newArgs.emplace_back(ast::MK::OptionalArg(symLoc, move(argName), ast::MK::Nil(symLoc)));
+        }
```

Comment:
Unless there's an `initialize` with a `sig`, I'm keeping the permissive signature that's implemented for `initialize` today. The optional positional arguments permit `new` invocations with positional arguments and keyword arguments, only validating that there are not too many positional arguments passed. But I'm moving it from `initialize` to `self.new` to leave room for a kwargs-only `initialize` sig to be provided, as discussed in another comment.